### PR TITLE
Changing resources created format

### LINKED ISSUE #136: [FEATURE] Resources created in state index and de-provisioning mapping
### Is your feature request related to a problem?
Right now we have a resources created field in the state index but there are a few ways on how we want to store this information.

### What solution would you like?
We can either map the ID of what was created to the type/name of the workflow step that created it or to the object name itself. 

Option 1 [Preferred]:

```
"resources_created": [
        {
            "type": "create_index_step",
            "id": "FobagIsBjjuFSw4Sfl3c"
        },
        {
            "type": "register_model",
            "id": "FobagIsBjjuFSw4Sfl3c"
        }
    ]
    
 ```
 Option 2:
 ```
"resources_created": [
        {
            "type": "index",
            "id": "FobagIsBjjuFSw4Sfl3c"
        },
        {
            "type": "model",
            "id": "FobagIsBjjuFSw4Sfl3c"
        }
    ]
    
 ```

I think it makes sense to map the created objected to the step that created it, in order to avoid adding another category of things like `model/index/agent`. 

However two things we need to consider are:
1. The order of the created objects in terms of de-provisioning order. Especially if some things depend on each other. For example do we need to delete the agent first before deleting the model that uses it. We can also potentially get this order from the use case template's nodes and edges themselves and delete in reverse order. 

2. How will we map each workflow step that created the resources to the workflow step that deletes the resources. We should probably have a way to match the create request to the equivalent delete request when we add workflow nodes/steps.

There could be a global delete map that looks something like this, this would be however another thing to maintain when adding new steps:
create_api:delete_api
```
 "deletion_map": {
            "create_index_step": "delete_index_step",
            "register_model": "delete_model",
        },
```

## Current Solution:
```
"resources_created": [
        {
            "workflow_step_name": "create_connector",
            "resource_id": "fZTmz4sBVvQ80zEwkC2z"
        }
    ]
    
 ```

### Additional Context

Right now we have some implementation of updating the resources created field in CreateConnectorStep. We should decide if we want to add more information here by fetching the resource name from some enum. 

Additionally we can abstract the method in CreateConnectorStep to be used by all steps easily as mentioned in this comment: https://github.com/opensearch-project/opensearch-ai-flow-framework/pull/148#discussion_r1393484371

### LINKED ISSUE #173: [FEATURE] Create an enum for Resources
### Is your feature request related to a problem?

To support the deprovisioning workflow we need a consistent set of Strings associated with resource types (e.g. `connector_id`).

Currently many of these are constant strings in the `common` package which is a good first step.  Some of them are hardcoded in the workflow steps which is not a good practice.

### What solution would you like?

Create an enum associated with Resources created where the enum name can be used (lowercased) for the JSON string key.

### PR DESCRIPTION
**Pull request by amitgalitz**
### Description

Added updates to resources_created section for register model, create index, create ingest pipeline and create connector steps.

Changed the format of creating a resource to: 
```
 {
                "_index": ".plugins-workflow-state",
                "_id": "RvEFF4wBKbeLfRazqAdh",
                "_score": 1.0,
                "_source": {
                    "workflow_id": "RvEFF4wBKbeLfRazqAdh",
                    "state": "PROVISIONING",
                    "provisioning_progress": "IN_PROGRESS",
                    "provision_start_time": 1701193321252,
                    "resources_created": [
                        {
                            "workflow_step_name": "create_connector",
                            "workflow_step_id": "workflow_step_1-amit-test",
                            "connector_id": "TfEGF4wBKbeLfRazCwcp"
                        },
                        {
                            "workflow_step_name": "register_remote_model",
                            "workflow_step_id": "workflow_step_2",
                            "model_id": "T_EGF4wBKbeLfRazCwdx"
                        }
                    ],
                    "provision_end_time": 1701193305145
                }
            }
```

### Issues Resolved
 resolves #173 
 resolves #136 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by dbwiddis — APPROVED** [review]
LGTM.   Note #215 will change the occurrences of `data.get(0)` to `currentNodeInputs`, whichever PR gets merged first will need to rebase for the other.

**Comment by dbwiddis** [comment]
Should this PR be against the `feature/agent_framework` branch?

**Comment by amitgalitz** [comment]
> Should this PR be against the `feature/agent_framework` branch?

Tbh not sure if we wanted to merge to feature branch and then port to main and 2.x or merge to main then 2.x then feature branch? @owaiskazi19 what was your preferred route?

**Comment by amitgalitz** [comment]
> If the feature's mostly standalone it can go to main and add the feature backport tag.
> 
> The problem is this one edits nearly every workflow step and will create a conflict with #215. Much easier to resolve if both PRs are on the same branch and you can rebase. Otherwise we'll end up with the conflicts when we eventually try to rebase main and 2.x with the feature branch.
> 
> I guess my position is:
> 
> * if [Improve workflow step execute API #215](https://github.com/opensearch-project/flow-framework/pull/215) gets merged first then target this against same branch and rebase
> * If this gets merged first, we need to make sure it gets backported to feature branch so I can rebase [Improve workflow step execute API #215](https://github.com/opensearch-project/flow-framework/pull/215)
> 
> Whatever we do, don't merge this to main and #215 to feature branch (without backport) because merge conflicts on cherry-picks are ugly

If you don't think rebasing will be too much work on your end I don't mind making sure I backport this immediately to feature branch, I do want to get this merged soon though and not leave it hanging cause it solves some update issues and has the base line for how to change resources that can easily be copied to new steps.

**Comment by owaiskazi19** [comment]
> > Should this PR be against the `feature/agent_framework` branch?
> 
> Tbh not sure if we wanted to merge to feature branch and then port to main and 2.x or merge to main then 2.x then feature branch? @owaiskazi19 what was your preferred route?

We can get this on `main` and then backport to `2.x` and `feature/agent_framework`

**Comment by owaiskazi19** [comment]
@amitgalitz Don't we need to add `resources_created` for `ModelGroupStep` or if any else workflow steps if we missed?

File: src/main/java/org/opensearch/flowframework/workflow/WorkflowStepFactory.java [hunk 1]
```diff
@@ -55,10 +55,10 @@ private void populateMap(
...
-        stepMap.put(CreateIngestPipelineStep.NAME, new CreateIngestPipelineStep(client));
-        stepMap.put(RegisterLocalModelStep.NAME, new RegisterLocalModelStep(mlClient));
-        stepMap.put(RegisterRemoteModelStep.NAME, new RegisterRemoteModelStep(mlClient));
+        stepMap.put(CreateIndexStep.NAME, new CreateIndexStep(clusterService, client, flowFrameworkIndicesHandler));
+        stepMap.put(CreateIngestPipelineStep.NAME, new CreateIngestPipelineStep(client, flowFrameworkIndicesHandler));
+        stepMap.put(RegisterLocalModelStep.NAME, new RegisterLocalModelStep(mlClient, flowFrameworkIndicesHandler));
+        stepMap.put(RegisterRemoteModelStep.NAME, new RegisterRemoteModelStep(mlClient, flowFrameworkIndicesHandler));
         stepMap.put(DeployModelStep.NAME, new DeployModelStep(mlClient));
         stepMap.put(CreateConnectorStep.NAME, new CreateConnectorStep(mlClient, flowFrameworkIndicesHandler));
         stepMap.put(ModelGroupStep.NAME, new ModelGroupStep(mlClient));
```
[the remark this answers is not in the document]
Comment (amitgalitz) [reply src/main/java/org/opensearch/flowframework/workflow/WorkflowStepFactory.java:64]: Didn't add it to deployModel and modelGroupStep. Model group is a miss by me, adding this now, for `DeployModelStep` we have no resource being created.

**Comment by amitgalitz** [comment]
Closing in favor of merging to the feature branch first: https://github.com/opensearch-project/flow-framework/pull/231

**Pull request closed without merging**